### ISSUE TITLE
Type params constrained to Record<string, ...> require type args to have index signatures

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#15161]
```ts
interface Foo {
    a: "hello" | "world";
    b: boolean
}

declare function takeFoo<T extends Record<string, string | boolean>>(x: T): T;

takeFoo<Foo>({
    a: "hello",
    b: true,
});
```

Expected: no error
Actual: 

```
Type 'Foo' does not satisfy the constraint 'Record<string, number | boolean>'.
  Index signature is missing in type 'Foo'.
```

**Comment by dmikis** [comment]
It's not immediately obvious why it's not a bug. Right now this breaks our project on 2.7.2 (worked perfectly on 2.5).

**Comment by RyanCavanaugh** [comment]
We're good enough at circular constraints now that this isn't needed. You can write
```ts
declare function takeFoo<T extends Record<keyof T, string | boolean>>(x: T): T;
```
